### DISCUSSION TITLE
Pathway to ARMI 1.0.0

### DISCUSSION THREAD

**Discussion by john-science** [terrapower/armi#1583] [General]
At time of writing, ARMI has a well-tested, stable codebase at version `0.3.0`.  But @keckler recently asked me what it would take to get ARMI to version `1.0.0`.  And that's a great question!

Obviously, I have a lot of ideas, but I think it's worth thinking through and getting some feedback from the userbase about the plan, before I re-publish the [ARMI Roadmap](https://github.com/terrapower/armi/wiki/Roadmap), which is getting a little old, and I would like to reframe it to this `1.0.0` question.

[Here](https://github.com/terrapower/armi/issues?q=is%3Aopen+is%3Aissue+label%3Aarchitecture) is an interesting place to look for ideas.

## v0.3.0

> Published 2024-02-02

## v0.4.0

> September-ish timeline

The focus of this release will be paying off our technical debt. Though some new features are in the works, most notably in assembly axial expansion and more detailed modeling of fuel pins.

### Must Have

* `HexGrid` work
  * [x] [Location of pins in tips-up hexagonal grid is off by 60 degrees](https://github.com/terrapower/armi/issues/1278)  (1-2 Days)
  * [x] [Incorrect pitch for tips up sub-block pin grid](https://github.com/terrapower/armi/issues/1284)  (1-3 days)
* [ ] [Reduce the trickiness of `ArmiObject.setNumberDensities()` by moving it to `Component`](https://github.com/terrapower/armi/issues/1374) (1-3 days)
* [ ] [`ArmiObject` has many, ultra-specific methods. Try to clean house.](https://github.com/terrapower/armi/issues/1375)
* [ ] [Move the lattice physics modules out of ARMI](https://github.com/terrapower/armi/tree/2e7304bae853541d2e0d5c4f1e37a738f28758eb/armi/physics/neutronics/latticePhysics)  (1-5 days)
* [ ] [Move the `crossSectionGroupManager` out of ARMI](https://github.com/terrapower/armi/blob/b9f1f1330b6b4354e2e27455402a7e641003a323/armi/physics/neutronics/crossSectionGroupManager.py)
* [ ] [Make EOC available in the DB](https://github.com/terrapower/armi/issues/1100) (1-5 days)
* [ ] [Axial Expansion with Multiple Pin Types](https://github.com/terrapower/armi/pull/1376)
* [ ] [Make the Doppler constant param VOLUME_INTEGRATED](https://github.com/terrapower/armi/pull/1659)
* [ ] [Inherit XS Settings when creating a new XS ID](https://github.com/terrapower/armi/pull/1653)
* [x] [Move dochelpers.py out of ARMI and into the /doc/ directory](https://github.com/terrapower/armi/issues/1661)

### Should Have

* [ ] Get code coverage over 90%.  (0-5 days)
* [ ] [Move Global Nuclides to something not global](https://github.com/terrapower/armi/discussions/1360) (1-10 days)
* [ ] [Remove the Operator from the Reactor](https://github.com/terrapower/armi/issues/830) (1-3 days)

### Nice-to Have

* [ ] [Redesign of migration scripts](https://github.com/terrapower/armi/issues/1657)
* [ ] [Improve High-Level ARMI Introduction](https://github.com/terrapower/armi/issues/1666)
* [ ] [Rename powerMultiplier to symmetryFactor](https://github.com/terrapower/armi/issues/1485) (1-2 days)

## v1.0.0

> No firm timeline.

This is the big one. This release is meant to be the line in the sand where ARMI's API is finally stable, and ARMI is feature rich without including too much science or engineering within ARMI.

### Features

* [ ] [Some support for ex-core systems](https://github.com/terrapower/armi/issues/262)
* [ ] materials package re-design

### Architecture

* [ ] [Finish removing all the globals we want to remove](https://github.com/terrapower/armi/discussions/1360)

### Quality Assurance

* [ ] [Choose a type system: type hints, or NumPy docs](https://github.com/terrapower/armi/discussions/1635) or [other](https://github.com/terrapower/armi/discussions/493)

### User Interface

* [ ] Improved user experience with preparing input files and analyzing output files is desired, possibly along the lines of e.g. [NEAMS Workbench](https://www.ornl.gov/onramp/neams-workbench) or with a web-based interface.

### Plugin Ecosystem

* [ ] We want to finish [the ARMI OpenMC plugin](https://github.com/terrapower/armicontrib-openmc).
* [ ] We'd love to see an ARMI input model of BEAVRS benchmark (#262).
* [ ] We'd like to see an ARMI plugin related to INL's Deep Lynx.

**Comment by john-science** [comment]
@sombrereau @jeffbaylor @keckler @drewj-usnctech @opotowsky Just FYI.  I am hoping to discuss this with all of you.

**Reply by dpham-materials** [reply]
I don't know if this belongs here, and I'm definitely not as familiar with the ARMI ecosystem, but the /materials/ folder could use some cleanup. From my newbie's glance, I can't tell what's being used and the essential properties for each type of material. The test coverage also looks a bit lacking.

**Reply by john-science** [reply]
Good point, David.  I will put "materials package rework" on version `1.0.0`, but hopefully we can get to it sooner than that.

**Comment by tp-pbritt-nuc** [comment]
Version 1.0.0 just denotes that it's past beta, correct?

I don't have any specific wants/needs. I would just say the "90% code coverage" is what I'd be looking for - make sure it works well enough to take to 1.0.0. As long as Pylot works, I'm not particular.

**Reply by john-science** [reply]
Well, let's not worry about terms like "alpha" and "beta" releases here.  Let's see this more as a way to Plan for what's Important.

Though I would like to knock all of ARMIs sharp edges off by the `1.0.0` release.

**Comment by drewj-usnctech** [comment]
Exciting! Okay, I have a frustrating "it depends" type answer: what does a 1.0.0 _mean_?

Does it imply stability in the framework and/or API to a reasonably degree such that there's some (implicit or explicit) compatibility promises? E.g., Applications that work with 1.0.0 will work with 1.0.X and 1.X.Y but not 2.X.Y?

Does it mean it's mostly got everything you'd want a reactor modeling framework to support? If that's the case, I would toss these items in the bucket
- #262 
- #228
- #1278
- #1284 

With these items, you can conceivably model the majority of reactor concepts in the wild. Or at least get close enough to be frustrated that ARMI gets close but not all the way

There is also this from the [ARMI docs](https://terrapower.github.io/armi/developer/tooling.html#releasing-a-new-version-of-armi)

> 
> We use the common major.minor.bump version scheme where a version string might look like 0.1.7, 1.0.0, or 12.3.123. Each number has a specific meaning:
> 
>     major - Revved for DOE/NRC-sanctioned release or at the end of a long development cycle.
> 
>     minor - Revved when we decide the code or our API has reached a stable point.
> 
>     bump - Revved every time we modify the API, and any other time we want.
> 

Type hints: yes please but maybe not required for 1.0.0 maybe before maybe later? There is a frustrating amount of double effort both doing type hints and numpydocs. This is a broader python pain point, but at least newer versions of sphinx are getting better at incorporating type hints?

**Reply by john-science** [reply]
I love it.  I'm really excited to pick some targets and work towards them. I am also excited to see what the user base wants most.

> major - Revved for DOE/NRC-sanctioned release or at the end of a long development cycle.

Yeah, this is an interesting point, because we are releasing ARMI to the NRC at version `0.3.0`. Mostly because I think there is still some work to do. I am not sure if I will updates the docs on that point or not. 

>  There is a frustrating amount of double effort both doing type hints and numpydocs.

Ugh.  I am not really interested in duplicating effort.  ARMI will go to "only type hints" or "only NumPy docs".  But I doubt we will support _both_, in `1.0.0`.

**Reply by opotowsky** [reply]
I'm team numpy docs for 1.0!

**Reply by albeanth** [reply]
I'm team type hints! 🥊

**Reply by tp-pbritt-nuc** [reply]
I'm team whatever-helps-me-navigate-the-type-nightmare-that-is-python....

**Reply by drewj-usnctech** [reply]
Made a poll for type hints - https://github.com/terrapower/armi/discussions/1635

**Comment by mgjarrett** [comment]
Here are a couple open PRs that should go in by v0.4.0:

https://github.com/terrapower/armi/pull/1659
https://github.com/terrapower/armi/pull/1653

**Reply by mgjarrett** [reply]
Also #1654, or at least part of it. It does two things:

1. fix a bug in "corners up" hex lattice pitch calculation
2. introduce a new way for specifying hex pitch in the blueprints: `latticePitch: {hex: 1.0}`

We might consider splitting that into two PRs; we would want (1) in ASAP because it's a bug fix. (2) is just an enhancement without an immediate motivating need, so it could wait if we want to avoid it for now.

**Comment by john-science** [comment]
The results of this conversation have been officially moved to the [ARMI Roadmap](https://github.com/terrapower/armi/wiki/Roadmap).
